docs/architecture/adr-001-node-key-refactoring.md

```diff
@@ -0,0 +1,101 @@
+
+The `orphans` are used to manage node removal in the current design and allow deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`, but there are not many use cases of `DeleteVersion`. There are two use cases:
+
+1. Rollback of the tree to a previous version
+2. Remove unnecessary old nodes
+
+## Decision
+
+- Use the sequenced integer ID as a node key like `bigendian(nodeKey)` format.
+- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field.
+- Remove the `orphans` from the tree.
+
+Theoretically, we can also remove the `version` field in the node structure but it leads to breaking the ics23 proof mechanism. We will revisit it later.
+
+New node structure
+
+```go
+type Node struct {
+	key           []byte
+	value         []byte
+	hash          []byte    // keep this field in the storage
+	nodeKey       int64     // new field, use as a node key
+	leftNodeKey   int64     // new field, need to store in the storage
+	rightNodeKey  int64     // new field, need to store in the storage
+	version       int64
+	size          int64
+	leftNode      *Node
+	rightNode     *Node
+	subtreeHeight int8
+	persisted     bool
+}
+```
+
+New tree structure
+
+```go
+type MutableTree struct {
+	*ImmutableTree                                    
+	lastSaved                *ImmutableTree
+	nonce                    int64                    // new field to track the current ID
+	versions                 map[int64]bool           
+	allRootLoaded            bool                     
+	unsavedFastNodeAdditions map[string]*fastnode.Node
+	unsavedFastNodeRemovals  map[string]interface{}   
+	ndb                      *nodeDB
+	skipFastStorageUpgrade   bool 
+
+	mtx sync.Mutex
+}
+```
+
+### Migration
+
+We can migrate nodes one by one by iterating the version.
+
+- Iterate the version in order, and get the root node for the specific version.
+- Iterate the tree and assign the `nodeKey` to nodes which the node version equals. 
+
+We will implement the `Import` functionality for the original version.
+
+### Pruning
+
+We introduce a new way to prune old versions.
+
+For example, when a user wants to prune the previous 500 versions every 1000 blocks
+- We assume the pruning is completed for `n`th version and the last nonce of `n`th version is `x`.
+- We iterate the tree from the `n+501`th root node and pick only nodes which the nodeKey is in `[(n+1)th version first nonce, (n+500)th version the last nonce]`.
+- We can delete missing nodes instantly or re-assign the nodeKey from `x+1` in order for those nodes. Re-assign should be done after stopping the node but it can lead to improving the data locality.
```

Comment:
I don't follow how pruning works, how do we know the node is not referenced by later versions so it's safe to delete?

Reply:
I was wondering that as well, but not sure if I just don't understand

Reply:
we traverse all nodes based on `n+501`th root, and pick the nodes which the nodeKey is greater than `x`. 
then sort those nodes by nodeKey and can find the missing nodes.

for example, x = 500, nodes= [501, 502, 504, 506 ...], then we can assume 503 and 505 nodes are missed, right?

Reply:
I see, use the n+501 tree as the gc root, and delete the missing ones. If a node is not referenced by n+501, it won't be referenced by later versions as well.

Reply:
exactly

Reply:
I think this needs to be made a lot clearer and I'm still having trouble understanding how this prunes all the nodes. How do we know that nodes <= x are all pruned properly? Can't nodes less than x still be referenced by later versions of the tree and be eligible for pruning after the nth version has been pruned?

Also, generally this seems like a somewhat roundabout way to do pruning that could become inefficient with large trees because it requires traversing the whole tree and then storing nodes in memory in a sorted list to find out which nodes in the sequence are missing right?

Why don't we just do reference counting?

Reply:
In my mind reference counting GC would make it simpler. It wouldn't require traversing all the nodes and would use very little memory. Has anyone explored that?

Reply:
The main idea of this story is to remove orphans, for me keeping orphans every time when updating the tree is very expensive. I'd like to separate pruning with main CRUD operations.

Reply:
> I guess the cost of reference counting on disk is non-trivial though.

It's just an integer. I think we wouldn't know without benchmarks. Maybe we can explore it in the future.

Anyway, removing orphans makes sense.

Reply:
traversing the whole tree is time-consuming. Shouldn't we try to avoid that?

Reply:
pruning is handled from the application now, if we can keep the same pruning style, I understand there will need to be minor changes, then it would be good. Later on we should explore native pruning to avoid  write